Aggregations throw ClassCastException with Predicates or PredicateBuilder

I'm filing this in response to this [StackOverflow question](http://stackoverflow.com/questions/29481508/hazelcast-aggregations-api-results-in-classcastexception-with-predicates).

I was refactoring some code to use the Aggregations API for counts, rather than the naive approach of building a list of matching entries and returning the results.

First attempt (using `PredicateBuilder`):
```java
 public static int reportCountforSource(String sourceId) 
  {
      EntryObject e = new PredicateBuilder().getEntryObject();
      Predicate<String, Report> predicate = e.get("sourceId").equal(sourceId);
      Supplier<String, Report, Object> supplier = Supplier.fromPredicate(predicate);
      Long count = reportMap.aggregate(supplier, Aggregations.count());

      return count.intValue();
  }  
```
Second attempt (using `Predicates`):

```java
  public static int reportCountforSource(String sourceId) 
  {
      @SuppressWarnings("unchecked")
      Predicate<String, Report> predicate = Predicates.equal("sourceId", sourceId);
      Supplier<String, Report, Object> supplier = Supplier.fromPredicate(predicate);
      Long count = reportMap.aggregate(supplier, Aggregations.count());

      return count.intValue();
  }    
```

Both of these methods result in a `ClassCastException`
```
Caused by: java.lang.ClassCastException: com.hazelcast.mapreduce.aggregation.impl.SupplierConsumingMapper$SimpleEntry cannot be cast to com.hazelcast.query.impl.QueryableEntry
    at com.hazelcast.query.Predicates$AbstractPredicate.readAttribute(Predicates.java:859) 
    at com.hazelcast.query.Predicates$EqualPredicate.apply(Predicates.java:779) 
    at com.hazelcast.mapreduce.aggregation.impl.PredicateSupplier.apply(PredicateSupplier.java:58) 
    at com.hazelcast.mapreduce.aggregation.impl.SupplierConsumingMapper.map(SupplierConsumingMapper.java:55)
    at com.hazelcast.mapreduce.impl.task.KeyValueSourceMappingPhase.executeMappingPhase(KeyValueSourceMappingPhase.java:49)
```
Third attempt (using lambda as `Predicate` interface implementation):
```java
 public static int reportCountforSource(String sourceId) 
  {
      Predicate<String, Report> predicate = (entry) ->  entry.getValue().getSourceId().equals(sourceId);
      Supplier<String, Report, Object> supplier = Supplier.fromPredicate(predicate);
      Long count = reportMap.aggregate(supplier, Aggregations.count());

      return count.intValue();
  }    
```

According to the response from @noctarius, this is a bug (or two).

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/SupplierConsumingMapper.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/SupplierConsumingMapper.java
@@ -16,16 +16,20 @@
 
 package com.hazelcast.mapreduce.aggregation.impl;
 
+import com.hazelcast.config.MapAttributeConfig;
 import com.hazelcast.mapreduce.Context;
 import com.hazelcast.mapreduce.Mapper;
 import com.hazelcast.mapreduce.aggregation.Supplier;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.query.impl.QueryableEntry;
+import com.hazelcast.query.impl.getters.Extractors;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 
 import java.io.IOException;
-import java.util.Map;
+import java.util.Collections;
 
 /**
  * The default mapper implementation for most (but not DistinctValues) aggregations.
@@ -85,20 +89,21 @@ class SupplierConsumingMapper<Key, ValueIn, ValueOut>
 
     /**
      * Internal implementation of an map entry with changeable value to prevent
-     * to much object allocation while supplying
+     * to much object allocation while supplying.
+     * Extends QueryableEntry in order to provide Extractable logic
      *
      * @param <K> key type
      * @param <V> value type
+     * @see com.hazelcast.query.impl.QueryableEntry
+     * @see com.hazelcast.query.impl.Extractable
      */
-    private static final class SimpleEntry<K, V>
-            implements Map.Entry<K, V> {
+    private static final class SimpleEntry<K, V> extends QueryableEntry<K, V> {
 
         private K key;
         private V value;
 
-        @Override
-        public K getKey() {
-            return key;
+        public SimpleEntry() {
+            this.extractors = new Extractors(Collections.<MapAttributeConfig>emptyList());
         }
 
         @Override
@@ -108,7 +113,32 @@ class SupplierConsumingMapper<Key, ValueIn, ValueOut>
 
         @Override
         public V setValue(V value) {
+            this.value = value;
+            return value;
+        }
+
+        @Override
+        public K getKey() {
+            return this.key;
+        }
+
+        @Override
+        public Data getKeyData() {
+            // not used in map-reduce
             throw new UnsupportedOperationException();
         }
+
+        @Override
+        public Data getValueData() {
+            // not used in map-reduce
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected Object getTargetObject(boolean key) {
+            return key ? this.key : this.value;
+        }
+
     }
+
 }

--- a/hazelcast/src/main/java/com/hazelcast/query/impl/Extractable.java
+++ b/hazelcast/src/main/java/com/hazelcast/query/impl/Extractable.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query.impl;
+
+import com.hazelcast.query.QueryException;
+
+/**
+ * Enables extracting attribute values and types and their type from an object, usually from an Entry
+ */
+public interface Extractable {
+
+    Object getAttributeValue(String attributeName) throws QueryException;
+
+    AttributeType getAttributeType(String attributeName) throws QueryException;
+
+}

--- a/hazelcast/src/main/java/com/hazelcast/query/impl/QueryableEntry.java
+++ b/hazelcast/src/main/java/com/hazelcast/query/impl/QueryableEntry.java
@@ -34,23 +34,28 @@ import static com.hazelcast.query.impl.TypeConverters.NULL_CONVERTER;
 
 /**
  * This abstract class contains methods related to Queryable Entry, which means searched an indexed by SQL query or predicate.
+ * <p/>
+ * If the object, which is used as the extraction target, is not of Data or Portable type the serializationService
+ * will not be touched at all.
  */
-public abstract class QueryableEntry implements Map.Entry {
+public abstract class QueryableEntry<K, V> implements Extractable, Map.Entry<K, V> {
 
     protected SerializationService serializationService;
     protected Extractors extractors;
 
+    @Override
     public Object getAttributeValue(String attributeName) throws QueryException {
         return extractAttributeValue(attributeName);
     }
 
+    @Override
     public AttributeType getAttributeType(String attributeName) throws QueryException {
         return extractAttributeType(attributeName);
     }
 
-    public abstract Object getValue();
+    public abstract V getValue();
 
-    public abstract Object getKey();
+    public abstract K getKey();
 
     public abstract Data getKeyData();
 
@@ -116,7 +121,7 @@ public abstract class QueryableEntry implements Map.Entry {
         if (KEY_ATTRIBUTE_NAME.value().equals(attributeName)) {
             return serializationService.toObject(key);
         } else if (THIS_ATTRIBUTE_NAME.value().equals(attributeName)) {
-            return serializationService.toObject(value);
+            return value instanceof Data ? serializationService.toObject(value) : value;
         }
         return null;
     }
@@ -143,7 +148,11 @@ public abstract class QueryableEntry implements Map.Entry {
             }
         }
 
-        Object targetObject = serializationService.toObject(target);
+        Object targetObject = target;
+        if (target instanceof Data) {
+            targetObject = serializationService.toObject(target);
+        }
+
         return extractors.extract(targetObject, attributeName);
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/query/impl/predicates/AbstractPredicate.java
+++ b/hazelcast/src/main/java/com/hazelcast/query/impl/predicates/AbstractPredicate.java
@@ -22,8 +22,8 @@ import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.query.Predicate;
 import com.hazelcast.query.QueryException;
 import com.hazelcast.query.impl.AttributeType;
+import com.hazelcast.query.impl.Extractable;
 import com.hazelcast.query.impl.IndexImpl;
-import com.hazelcast.query.impl.QueryableEntry;
 import com.hazelcast.query.impl.getters.MultiResult;
 
 import java.io.IOException;
@@ -100,7 +100,7 @@ public abstract class AbstractPredicate
                 // Returning unconverted value is an optimization since the given value will be compared with null.
                 return givenAttributeValue;
             }
-            type = ((QueryableEntry) entry).getAttributeType(attributeName);
+            type = ((Extractable) entry).getAttributeType(attributeName);
             attributeType = type;
         }
 
@@ -126,8 +126,8 @@ public abstract class AbstractPredicate
     }
 
     protected Object readAttributeValue(Map.Entry entry) {
-        QueryableEntry queryableEntry = (QueryableEntry) entry;
-        Object attributeValue = queryableEntry.getAttributeValue(attributeName);
+        Extractable extractable = (Extractable) entry;
+        Object attributeValue = extractable.getAttributeValue(attributeName);
         return convertEnumValue(attributeValue);
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/MapBaseAggregationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/MapBaseAggregationTest.java
@@ -19,6 +19,7 @@ package com.hazelcast.mapreduce.aggregation;
 import com.hazelcast.core.IMap;
 import com.hazelcast.mapreduce.KeyPredicate;
 import com.hazelcast.query.Predicate;
+import com.hazelcast.query.Predicates;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
@@ -26,6 +27,7 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.io.Serializable;
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.Map;
@@ -116,6 +118,62 @@ public class MapBaseAggregationTest
 
     @Test
     public void testPredicateAggregation()
+            throws Exception {
+
+        String mapName = randomMapName();
+        IMap<Integer, Integer> map = HAZELCAST_INSTANCE.getMap(mapName);
+
+        Integer[] values = buildPlainValues(new ValueProvider<Integer>() {
+            @Override
+            public Integer provideRandom(Random random) {
+                return random(1000, 2000);
+            }
+        }, Integer.class);
+
+        for (int i = 0; i < values.length; i++) {
+            map.put(i, values[i]);
+        }
+
+        Predicate<Integer, Integer> predicate = Predicates.greaterEqual("this", 1000);
+        Supplier<Integer, Integer, Object> supplier = Supplier.fromPredicate(predicate);
+        Aggregation<Integer, Object, Long> aggregation = Aggregations.count();
+        long count = map.aggregate(supplier, aggregation);
+        assertEquals(values.length, count);
+    }
+
+    private static class Person implements Serializable {
+        Integer age;
+    }
+
+    @Test
+    public void testPredicateAggregation_customExtraction()
+            throws Exception {
+
+        String mapName = randomMapName();
+        IMap<Integer, Person> map = HAZELCAST_INSTANCE.getMap(mapName);
+
+        Person[] values = buildPlainValues(new ValueProvider<Person>() {
+            @Override
+            public Person provideRandom(Random random) {
+                Person person = new Person();
+                person.age = random(1000, 2000);
+                return person;
+            }
+        }, Person.class);
+
+        for (int i = 0; i < values.length; i++) {
+            map.put(i, values[i]);
+        }
+
+        Predicate<Integer, Person> predicate = Predicates.greaterEqual("age", 1000);
+        Supplier<Integer, Person, Object> supplier = Supplier.fromPredicate(predicate);
+        Aggregation<Integer, Object, Long> aggregation = Aggregations.count();
+        long count = map.aggregate(supplier, aggregation);
+        assertEquals(values.length, count);
+    }
+
+    @Test
+    public void testCustomPredicateAggregation()
             throws Exception {
 
         String mapName = randomMapName();

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/MultiMapBaseAggregationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/MultiMapBaseAggregationTest.java
@@ -19,6 +19,7 @@ package com.hazelcast.mapreduce.aggregation;
 import com.hazelcast.core.MultiMap;
 import com.hazelcast.mapreduce.KeyPredicate;
 import com.hazelcast.query.Predicate;
+import com.hazelcast.query.Predicates;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
@@ -26,6 +27,7 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.io.Serializable;
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.Map;
@@ -116,6 +118,62 @@ public class MultiMapBaseAggregationTest
 
     @Test
     public void testPredicateAggregation()
+            throws Exception {
+
+        String mapName = randomMapName();
+        MultiMap<Integer, Integer> map = HAZELCAST_INSTANCE.getMultiMap(mapName);
+
+        Integer[] values = buildPlainValues(new ValueProvider<Integer>() {
+            @Override
+            public Integer provideRandom(Random random) {
+                return random(1000, 2000);
+            }
+        }, Integer.class);
+
+        for (int i = 0; i < values.length; i++) {
+            map.put(i, values[i]);
+        }
+
+        Predicate<Integer, Integer> predicate = Predicates.greaterEqual("this", 1000);
+        Supplier<Integer, Integer, Object> supplier = Supplier.fromPredicate(predicate);
+        Aggregation<Integer, Object, Long> aggregation = Aggregations.count();
+        long count = map.aggregate(supplier, aggregation);
+        assertEquals(values.length, count);
+    }
+
+    private static class Person implements Serializable {
+        Integer age;
+    }
+
+    @Test
+    public void testPredicateAggregation_customExtraction()
+            throws Exception {
+
+        String mapName = randomMapName();
+        MultiMap<Integer, Person> map = HAZELCAST_INSTANCE.getMultiMap(mapName);
+
+        Person[] values = buildPlainValues(new ValueProvider<Person>() {
+            @Override
+            public Person provideRandom(Random random) {
+                Person person = new Person();
+                person.age = random(1000, 2000);
+                return person;
+            }
+        }, Person.class);
+
+        for (int i = 0; i < values.length; i++) {
+            map.put(i, values[i]);
+        }
+
+        Predicate<Integer, Person> predicate = Predicates.greaterEqual("age", 1000);
+        Supplier<Integer, Person, Object> supplier = Supplier.fromPredicate(predicate);
+        Aggregation<Integer, Object, Long> aggregation = Aggregations.count();
+        long count = map.aggregate(supplier, aggregation);
+        assertEquals(values.length, count);
+    }
+
+    @Test
+    public void testCustomPredicateAggregation()
             throws Exception {
 
         String mapName = randomMapName();

